Prerender heatmap in partial

hlfan opened the pull request:
Ditching the ever-increasing amount of code required to accommodate CalHeatmap like in #5906 and going for a custom responsive solution.
Fixes #5905 and closes #5906. Also fixes #5810.
This also makes the history links of the heatmap tab-navigatable, even before hovering over each link.

HolgerJeromin commented:
Nice. 🥳
Can you add a screenshot when ready?

hlfan commented:
default (ISO 8601):
![localhost_3000_user_hltest](https://github.com/user-attachments/assets/9ff7eb00-cf4a-4f1e-84c3-35cd9c05fb82)

"ar-AF" locale:
![localhost_3000_user_hltest (1)](https://github.com/user-attachments/assets/4760ee77-8a5c-456f-b50e-8da321576c90)

"dv-MV" locale:
![localhost_3000_user_hltest (2)](https://github.com/user-attachments/assets/e3903509-4de8-4e48-b221-274c73d95c9a)

kcne commented:
Even though I spent quite a bit of time trying to make our custom configuration work alongside cal-heatmap, I have to say - this approach makes a lot more sense. Modifying the library ended up requiring more code than writing the config itself.

I really like that this removes the third-party dependency and results in a much cleaner, more maintainable, and self-contained solution.

Just a quick question: what border-radius did you end up using for the squares? It seems to vary slightly across the three screenshots. Also, slightly increasing the spacing between columns and rows could help make the heatmap look a bit more balanced and visually appealing.

Thank you.

tomhughes commented:
With this approach is the javascript actually doing anything that couldn't be done in the rails rendering on the server?

hlfan commented:
For now, I've stopped just short of CLDR-dependent stuff using `Intl.locale`. But the [TwitterCldr](https://github.com/twitter/twitter-cldr-rb) gem seems to have the necessary information to move everything to Ruby.

tomhughes commented on app/assets/javascripts/heatmap.js:
```diff
@@ -1,91 +1,62 @@
+      .css("grid-area", getWeekdayRow(date.getUTCDay()) + " / " + weekColumn)
+      .attr("aria-label", tooltipText)
+      .tooltip({
+        title: tooltipText,
```
Currently the tooltips are prone to line wrapping which doesn't look very good so I'd suggest adding a custom class here like:

```js
customClass: "heatmap-tooltip"
```

and then a CSS rule to allow wider tooltips:

```css
.heatmap-tooltip {
  --bs-tooltip-max-width: 300px;
}
```

hlfan commented:
I've now moved the bootstrap styling to the CSS, saving about a dozen+ kB in the page response size.
Also `data-month` now counts beyond 12 so I don't have to deal with jQuery selecting both labels of the same month.
And I've fixed the month labels when the calendar reaches into a month by adding more padding for the month label generation.

kcne commented:
I've noticed that the border radius of the day cells changes between desktop and mobile — it looks slightly rounded on desktop but becomes almost pill-shaped on mobile. If this isn’t intentional, we might want to set a fixed border-radius to ensure visual consistency across breakpoints.

Desktop:
<img width="136" alt="Screenshot 2025-05-15 at 10 09 28" src="https://github.com/user-attachments/assets/18e1278a-6a56-4f3f-accf-604d24ba604e" />

Mobile:
<img width="90" alt="Screenshot 2025-05-15 at 10 09 35" src="https://github.com/user-attachments/assets/3a0de4c4-3d66-4632-812e-6ccb2faa5408" />


On macOS using firefox, when scrolling horizontally, the scrollbar overlaps with the heatmap area. It might be a good idea to add some padding at the bottom of the container to prevent this overlap and improve the scrolling experience. 

<img width="531" alt="Screenshot 2025-05-15 at 10 05 17" src="https://github.com/user-attachments/assets/d7a3f0f2-d434-4429-bc52-d6c3a78fe846" />

It might also be a good idea to apply light/dark variants for tooltips based on the current theme to keep visual consistency.

hlfan commented:
> we might want to set a fixed border-radius

This is what is currently being applied with Bootstrap's `rounded-1`, but I agree a relative border-radius looks more consistent.

> the scrollbar overlaps with the heatmap area

I've added `mb-1` as a compromise between thick overlaid and classic scrollbars, as I can't get `scrollbar-gutter` to produce the desired effect.

> apply light/dark variants for tooltips

That's what Bootstrap already does, although to prevent the tooltip "merging" with the background and for better accessibility, dark tooltips are used in light mode and light ones in dark mode.

tomhughes commented on app/assets/stylesheets/common.scss:
```diff
@@ -960,6 +960,33 @@ img.trace_image {
+}
+
+.tooltip.wide {
+	--bs-tooltip-max-width: none;
```
You've got a tab here as the indent.

hlfan replied:
Not again me not checking the style from the chrome dev tools...

hlfan replied:
Ok I've made myself a pre-commit hook that checks that this doesn't happen again.

tomhughes reviewed (approved):
Thanks. I think this is good now.

verhovsky commented:
I don't like the green.

hlfan commented:
Could this be a new edition of #5802 but for days? And caching is probably also involved.

Which green would you prefer (and for which color scheme)?

In general it would be better to open a new issue for that though.

verhovsky commented:
> Which green

The same one as github

<img width="147" alt="Screenshot 2025-05-16 at 15 09 05" src="https://github.com/user-attachments/assets/52e87e2d-6399-48df-b8f4-623eb613fd61" />

and I think their implementation of just 5 color buckets is easier to read than a continuous gradient.

Also, it would be cool if there was a "See more" button that would load the entire contribution graph since account creation, like github lets you click through the years

<img width="208" alt="Screenshot 2025-05-16 at 15 11 07" src="https://github.com/user-attachments/assets/a312a448-8fbf-4a0e-8f9c-8a084cb789e8" />

hlfan commented:
The yearly history has already been mentioned in #5829, which should IMO be a precondition for that.

And anything involving different light and dark color palletes requires watching the color scheme media query again, which I really want to avoid.

git-tgo commented:
nenad-vujicic > ... looks now (Edge, Chrome ...

I can partly agree.

  | logged in | logged out
-- | -- | --
FF | ok | ok
Edge | not ok | ok
Chrome | not ok | ok

tomhughes commented:
Looks OK to me in both Firefox and Chromium on linux whether or not I am logged in.

hlfan commented:
I think I've found the culprit:
https://github.com/openstreetmap/openstreetmap-website/blob/32f23461fc37f93ba7fcddd901ed156d544c575e/app/assets/javascripts/heatmap.js#L23
getUTCDay is [0,6], weekInfo.firstDay is [1,7]

verhovsky commented:
> just 5 color buckets is easier to read than a continuous gradient

for example I can't see any of my contributions because in the last few months I've had days with a lot more contributions

<img width="613" alt="Screenshot 2025-05-17 at 11 14 13" src="https://github.com/user-attachments/assets/0c3834ac-8998-4097-9adf-c348c6e5ffff" />
